src/Orleans/Utils/IKeyedServiceDescriptor.cs

```diff
@@ -0,0 +1,85 @@
+﻿
+using System;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans
+{
+    public static class KeyedServiceDescriptorExtensions
+    {
+        /// <summary>
+        /// Add supported for keyed services to a service collection
+        /// </summary>
+        public static IServiceCollection SupportKeyedServices(this IServiceCollection collection)
+        {
+            return collection.AddSingleton(typeof(IKeyedServiceCollection<,>), typeof(KeyedServiceCollection<,>));
+        }
+
+        /// <summary>
+        /// Associates a key with a specialization of a service type.
+        /// </summary>
+        public static IServiceCollection WithKey<TKey, TService, TSpecialization>(this IServiceCollection collection, TKey key)
```

Comment:
I agree that the fluent approach doesn't work for creating non-trivial registrations in IServiceCollection. It would be nice to get an Autofac-like API, but this is what we've got.